pkg/scheduler/schedule_one.go

```diff
@@ -283,6 +283,15 @@ func (sched *Scheduler) bindingCycle(
 		return status
 	}
 
+	// Any failures after this point cannot lead to the Pod being considered unshcedulable.
+	// i.e., They could return error only when some unexpected temporal errors happen. (e.g., kube-apiserver failure)
+	// So, if Pods are failed at one of them and going to the scheduling queue, the scheduling queue always pushes them to activeQ/backoffQ, instead of the unschedulable pod pool.
+	// This is because such Pods won't be schedulable by any certain cluster events, and thus scheduling hints aren't useful to determine when to push those Pods to activeQ/backoffQ.
```

Comment:
On second thoughts, this PR can be merged without https://github.com/kubernetes/kubernetes/pull/119607.

Without https://github.com/kubernetes/kubernetes/pull/119607, this kind of scenario could happen:
1. Pod goes through the scheduling cycle successfully. The binding cycle is started at goroutineA.
2. [goroutineA] Done() is called right before binding. This Pod is removed from `inFlightPods`.
3. The Pod is updated by someone and, due to https://github.com/kubernetes/kubernetes/issues/118226, this Pod enters the scheduling cycle while the binding cycle of the same Pod is still on-going. The Pod is again added to `inFlightPods` in Pop().
4. [goroutineA] Binding fails. This Pod will be moved back to the scheduling queue.
5. [goroutineA] `AddUnschedulableIfNotPresent` is called to push this Pod into the scheduling queue. And Done() is again called in `AddUnschedulableIfNotPresent`. This Done() removes `inFlightPods` added at (3).
6. The scheduling cycle started at (3) is rejected by some plugins. This Pod is moved back to the scheduling queue. `AddUnschedulableIfNotPresent` is called again, but this Pod isn't registered in `inFlightPods` (due to (5)). 

But, (6) doesn't matter actually because the Pod that came back to the queue at (5) should be still in the queue, and this coming back Pod at (6) is just ignored ([ref](https://github.com/sanposhiho/kubernetes/blob/3e438c54a77ccf7b9e0aff924eb977031099cc2f/pkg/scheduler/internal/queue/scheduling_queue.go#L766-L775)) 

In the end, people would see an error log from the scheduler ([ref](https://github.com/sanposhiho/kubernetes/blob/3e438c54a77ccf7b9e0aff924eb977031099cc2f/pkg/scheduler/internal/queue/scheduling_queue.go#L766-L775)), but that's it. (Also, seemingly we can say this scenario is rare.)
I cannot imagine any worse scenario that could make Pod being stuck in the unschedQ by merging this PR without #119607.

Reply:
isn't done already called at handleSchedulingFailure, which gets invoked in case of a binding error? Why do we need it here?

https://github.com/kubernetes/kubernetes/blob/85d903e27ee92e999839e674577a3e79400835a0/pkg/scheduler/schedule_one.go#L985

Reply:
Yes, it's already called, but the point is calling it before Bind, which contains API call, so that we can reduce the number of events stored in inFlightEvents (https://github.com/kubernetes/kubernetes/issues/120622).

[PR description:](https://github.com/kubernetes/kubernetes/pull/120586)
> All cluster events during scheduling are piled up in [inFlightEvents](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L187).
...
By calling Done() as soon as possible, we can reduce the memory consumed in inFlightEvents.